### ISSUE TITLE
Allow Passing of Established SQL Connections

### ISSUE THREAD

**Issue by joshua-oss** [opendp/smartnoise-sdk#325]
Our current design tightly couples information that we need for differential privacy (such as NameCompare) with engine-proprietary connection information that is used to establish a database connection.  This limits callers to connection types that we support (currently only username/password to an IP host and port).  This limitation is unnatural and unnecessary, since callers are accustomed to managing connections outside the scope of calling APIs.

The proposal is to update the API to allow callers to pass in established connections.  This has several advantages:
* Works seamlessly with external middleware such as gateways and load balancers
* Works seamlessly with more robust authentication techniques like certificate-based authentication
* Minimizes requirement for caller to learn new connection API
* Allows caller to externally manage transactions, including nested and distributed
* Facilitates transparent attach of engine-specific behavior we need

# Code Samples

The following code samples use a static factory method.  This is not the only possible design.

## Per-Engine Factory

```
import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

reader = PostgresReader.from_connection(conn)

priv_reader = PrivateReader(reader, meta)
```

This example improves on the current design, because the caller can use proprietary connection properties that apply only to the Postgres implementation of DP-API.  The `PostgresReader` uses the supplied connection, and simply serves as an access point for the `PostgresNameCompare` and `PostgresSerializer`.

Implementation cost is low.

## Generic Factory

We already have dynamic switching of `NameCompare` functionality based on the `engine` property in metadata.  This was a requirement, because we need to be able to parse and load symbols without an active connection.

```
import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

reader = SqlReader.from_connection(conn, "postgres")

priv_reader = PrivateReader(reader, meta)
```

The advantage of this approach is that the caller doesn't need to import classes or memorize the class names.  All connections are wrapped directly from SqlReader object.  

A disadvantage is that callers may use mismatching engine names (e.g. pass a SQL Server connection to the facory, while asking for "postgres" overrides).  This could be desirable, for example, if a caller has an unsupported 3rd-party database connection that is known to be compatible with an engine we support.

Implementation cost for this would be low.

## Generic Factory, Simplified

```
import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

reader = SqlReader.from_connection(conn)

priv_reader = PrivateReader(reader, meta)
```

We could simplify the call by eliminating the engine parameter or making it optional.  In this case, the factory constructor would query the connection for version information, and instantiate the appropriate engine-specific base class.  This approach is simpler for the caller, and ensures that the name compare and serialization behavior match the engine being used.

We may want to keep the engine parameter as optional, in case there are unknown 3rd party database which don't allow the engine to be determined from the connection.  This might be the case with specialized middleware.  If we keep the engine as optional, we could still issue a warning if the connection is made with a user-specified engine that doesn't match any engine version we can detect from the connection.

This would be a modest implementation effort.

## Generic Factory, Transparent

```
import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

priv_reader = PrivateReader(conn, meta)
```

A further simplification would automate the factory step in the `PrivateReader` constructor.  The `PrivateReader` would call `SqlReader.from_connection(conn)` transparently under the covers.  Existing calling patterns could be supported, in cases where caller needs to override some of the property attachment.  But this pattern should support the vast majority of scenarios.

Implementation is low cost on top of the "Simplified" proposal above.

Note that it might be more consistent to design this like `PrivateReader.from_connection(conn)`.  But we need to pass other parameters (metadata, epsilon, etc.), which doesn't match the intent of the factory method on the base classes.  Maybe we could add the params on the static method, like `PrivateReader.from_connection(conn, meta)`, and then slowly deprecate the constructor methods on all base classes and `PrivateReader`.

Implementation would be low cost

## Curator-Provided Metadata

We have already moved the metadata param to come second in constructors, to support cases where metadata will be tightly-bound to the database.  The platform could support maintaining the metadata in a special table name, and metadata would be loaded automatically.  This reduces opportunity for callers to mangle or override curator-specified constraints.  In this case, the calling pattern would be further simplified:

```
import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

priv_reader = PrivateReader(conn)
```

# Reference

The system is designed to work with any DB-API Connection object, but we create them ourselves, rather than allowing the caller to connect.

DB-API Connection specification:  https://www.python.org/dev/peps/pep-0249/#connection
PyODBC Connection (inherits from DB-API Connection): https://github.com/mkleehammer/pyodbc/wiki/Connection

**Comment by eedeleon** [comment]
This is a good idea. I think the best benefit here is rather than failing at creation time, the db connection is made and passed to the class.

I think the GenericFactory is the better of the options because it will reduce # classes and passing the engine name is something that can be verified.

It may be possible to default the engine, but I will let that be decided by someone who has more context.

import pyscopg2
conn = psycopg2.connect(dbname='PUMS', user='pums', sslmode='verify-full')

reader = SqlReader.from_connection(conn, "postgres")

priv_reader = PrivateReader(reader, meta)


I think SqlReader(conn)  # should be possible longer term, we may have to overload the current parameters or take a breaking change.

**Comment by joshua-oss** [comment]
Sounds good.  That keeps everything consistent with how we do NameCompare right now.
 
 I'm thinking the the implementation will pass through to an equivalent factory method on the underlying reader class, like:
 
 ```
 class SqlReader:
 	def from_connection(conn, engine):
		# look up concrete class (similar to NameCompare)
		return PostgresReader.from_connection(conn)
```

An alternative option would be to extend the constructors of underlying readers,like:

 ```
 class SqlReader:
 	def from_connection(conn, engine):
		# look up concrete class (similar to NameCompare)
		return PostgresReader(connection=conn)
```

An advantage of the first approach is that it keeps the legacy constructor-based approach in one code path where we can insert a warning and deprecate.  And we can immediately switch guidance for concrete reader implementors to implement only the static factory method.  The idea being that concrete reader implementations should be stateless (other than the connection), and just serve as a holder for engine-specific functionality like identity comparison.  The second approach could be better if we want to allow people to keep calling the legacy pattern optionally, while making it more flexible.

**Comment by eedeleon** [comment]
The main issue with having a polymorphic return here is the definition of the class will be circular since SqlReader would depend on PostgresReader for that factory method

**Comment by joshua-oss** [comment]
Could we solve that with the same approach we use for NameCompare?

https://github.com/opendifferentialprivacy/smartnoise-sdk/blob/master/sdk/opendp/smartnoise/sql/reader/base.py#L38

The NameCompare concrete classes are stored in the same files as the PostgresReader, PandasReader, etc.  Essentially, each concrete reader is just a bucket to hold the concrete PostgresNameCompare and PostgresSerializer, and we've already implemented the dictionary to load the custom NameCompare classes.

**Comment by joshua-oss** [comment]
Fixed in #375.  Syntax is `SqlReader.from_connection`
